daisy/disk.go

```diff
@@ -50,9 +57,107 @@ func (dm *diskMap) deleteFn(r *resource) error {
+		return Errorf("cannot attach disk to instance %q, does not exist", iName)
+	}
+	// Iterate over disk's attachments. Check for concurrent conflicts.
+	// Step s is concurrent with other attachments if the attachment detacher == nil
+	// or s does not depend on the detacher.
+	// If this is a repeat attachment (same disk and instance already attached), do nothing and return.
+	for attI, att := range dm.attachments[d] {
+		// Is this a concurrent attachment?
+		if att.detacher == nil || !s.depends(att.detacher) {
+			if attI == i {
+				return nil // this is a repeat attachment to the same instance -- does nothing
+			} else if strIn(diskModeRW, []string{mode, att.mode}) {
+				// Can't have concurrent attachment in RW mode.
+				return Errorf(
+					"disk attachment conflict for disk %q: attached to instances %q (%s) and %q (%s)",
+					dName, i.real, mode, attI.real, att.mode)
+			}
+		}
+	}
+
+	var im map[*resource]*diskAttachment
+	if im, ok = dm.attachments[d]; !ok {
+		im = map[*resource]*diskAttachment{}
+		dm.attachments[d] = im
+	}
+	im[i] = &diskAttachment{mode: mode, attacher: s}
 	return nil
 }
 
-func (dm *diskMap) registerDetachment(dName, iName, mode string, s *Step) error {
+// detachHelper marks s as the detacher between d and i.
+// Returns an error if d and i aren't attached
+// or if the detacher doesn't depend on the attacher.
+func (dm *diskMap) detachHelper(d, i *resource, s *Step) error {
+	if dm.testDetachHelper != nil {
+		return dm.testDetachHelper(d, i, s)
+	}
+	var att *diskAttachment
+	var im map[*resource]*diskAttachment
+	var ok bool
+	if im, ok = dm.attachments[d]; !ok {
+		return Errorf("not attached")
+	}
+	if att, ok = im[i]; !ok || att.detacher != nil {
+		return Errorf("not attached")
+	} else if !s.depends(att.attacher) {
+		return Errorf("detacher %q does not depend on attacher %q", s.name, att.attacher.name)
+	}
+	att.detacher = s
 	return nil
 }
+
+// registerDetachment marks s as the detacher for the dName disk and iName instance.
+// Returns an error if dName or iName don't exist
+// or if detachHelper returns an error.
+func (dm *diskMap) registerDetachment(dName, iName string, s *Step) error {
+	dm.mx.Lock()
+	defer dm.mx.Unlock()
+	var d, i *resource
+	var ok bool
+	if d, ok = dm.m[dName]; !ok {
+		return Errorf("cannot detach disk %q, does not exist", dName)
+	}
+	if i, ok = instances[dm.w].get(iName); !ok {
+		return Errorf("cannot detach disk from instance %q, does not exist", iName)
+	}
+
+	if err := dm.detachHelper(d, i, s); err != nil {
+		return Errorf("cannot detach disk %q from instance %q: %v", dName, iName, err)
+	}
+	return nil
+}
+
+// registerAllDetachments marks s as the detacher for all disks attached to the iName instance.
+// Returns an error if iName does not exist
+// or if detachHelper returns an error.
+func (dm *diskMap) registerAllDetachments(iName string, s *Step) error {
+	dm.mx.Lock()
+	defer dm.mx.Unlock()
+
+	var d, i *resource
```

Comment:
These declarations don't need to happen here, this is cleaner:

i, ok := instances[dm.w].get(iName)
if !ok {...}

var err Errors
for d := range dm.attachments {...}